test/unit-tests/Spectacle/Syntax/Logic.hs

```diff
@@ -0,0 +1,92 @@
+
+module Spectacle.Syntax.Logic
+  ( tests,
+  )
+where
+
+import Control.Monad
+import Data.Function ((&))
+import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import Test.Tasty
+import Test.Tasty.Hedgehog
+
+import Language.Spectacle.Lang
+import Language.Spectacle.Syntax.Error
+import Language.Spectacle.Syntax.NonDet
+import Language.Spectacle.Syntax.Logic
+import Language.Spectacle.Exception.RuntimeException
+
+-- -------------------------------------------------------------------------------------------------
+
+prop_negationInvolute :: Property
+prop_negationInvolute = property do
+  let negated = runLogical (complement (complement (pure True)))
+  case negated of
+    Left exc -> do
+      footnoteShow exc
+      failure
+    Right x -> x === [True]
+
+evensForall ::
+  Members '[Logic, NonDet, Error RuntimeException] effs =>
+  [Int] ->
+  Lang ctx effs Int
+evensForall xs = forall xs (pure . even)
```

Comment:
Does this `forall` function conflict with the `forall` keyword when certain language extensions are enabled (e.g. `ExplicitForall`, `RankNTypes`, `ScopedTypeVariables`)?

Reply:
Nope, `forall` exists in the function namespace and GHC doesn't look for quantifier keywords there.